sky/dag.py

```diff
@@ -2,74 +2,236 @@
+        >>>     task1 >> task2
     """
 
-    def __init__(self) -> None:
+    def __init__(self, name: Optional[str] = None) -> None:
+        """Initialize a new DAG.
+
+        Args:
+            name: Optional name for the DAG.
+        """
+
+        self.name = name
         self.tasks: List['task.Task'] = []
-        import networkx as nx  # pylint: disable=import-outside-toplevel
+        self._task_name_lookup: Dict[str, 'task.Task'] = {}
+        self.dependencies: Dict['task.Task', Set['task.Task']] = {}
 
         self.graph = nx.DiGraph()
-        self.name: Optional[str] = None
+
+    def _get_task(self, task_or_name: TaskOrName) -> 'task.Task':
+        """Get a task object from a task or its name.
+
+        Args:
+            task_or_name: Either a Task object or the name of a task.
+
+        Returns:
+            The Task object.
+
+        Raises:
+            ValueError: If the task name is not found in the DAG.
+        """
+        if not isinstance(task_or_name, str):
+            return task_or_name
+        name = task_or_name
+        if name not in self._task_name_lookup:
+            raise ValueError(f'Task {name} not found in DAG')
+        return self._task_name_lookup[name]
 
     def add(self, task: 'task.Task') -> None:
+        """Add a task to the DAG.
+
+        Args:
+            task: The Task object to add.
+
+        Raises:
+            ValueError: If the task already exists in the DAG or if its name
+            is already used.
+        """
+        if task.name is None:
+            task.name = serve_utils.generate_task_name(task)
+        if task in self.tasks:
+            raise ValueError(f'Task {task.name} already exists in the DAG.')
+        if task.name in self._task_name_lookup:
+            raise ValueError(
+                f'Task name "{task.name}" is already used in the DAG.')
         self.graph.add_node(task)
         self.tasks.append(task)
+        self._task_name_lookup[task.name] = task
+
+    def remove(self, task: Union['task.Task', str]) -> None:
+        """Remove a task from the DAG.
+
+        Args:
+            task: The Task object or name of the task to remove.
 
-    def remove(self, task: 'task.Task') -> None:
+        Raises:
+            ValueError: If the task is still being depended on by other tasks.
+        """
+        task = self._get_task(task)
+
+        dependents = list(self.graph.successors(task))
+        for dependent in dependents:
+            self.remove_dependency(dependent, task)
+
+        # TODO(andy): Stuck by optimizer's wrong way to remove dummy sources
+        # and sink nodes.
+        # if dependents:
+        #     dependent_names = ', '.join([dep.name for dep in dependents])
+        #     raise ValueError(f'Task {task.name} is still being depended on '
+        #                      f'by tasks {dependent_names!r}. Try to remove '
+        #                      'the dependencies first.')
```

Comment:
This TODO exists because our optimizer's usage is not strictly following the DAG rules. It attempts to remove sink/dummy nodes while they're still being depended on by other nodes. 

I believe it's better to keep this TODO for now. This check is actually beneficial for users, and we should address this issue when we refactor the optimizer. This way, we can solve both the optimizer's behavior and this check simultaneously.

Reply:
Sounds good! Could you help file an issue to keep track of this? Thanks!